server/src/main/java/com/hedera/block/server/verification/hasher/ConcurrentStreamingTreeHasher.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.block.server.verification.hasher;
+
+import static com.hedera.block.server.verification.hasher.CommonUtils.noThrowSha384HashOf;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.nio.ByteBuffer;
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * A {@link StreamingTreeHasher} that computes the root hash of a perfect binary Merkle tree of {@link Bytes} leaves
+ * using a concurrent algorithm that hashes leaves in parallel and combines the resulting hashes in parallel.
+ * <p>
+ * <b>Important:</b> This class is not thread-safe, and client code must not make concurrent calls to
```

Comment:
How can we be sure that this will be the case ? Do we wanna consider using some synchronization primitives ?

Reply:
This `StreamingTreeHasher` implementation are being imported from services `as-is`